components/brave_rewards/resources/extension/brave_rewards/background/reducers/promotion_reducer.ts

```diff
@@ -0,0 +1,201 @@
+        break
+      }
+
+      const hint = payload.properties.hint
+      const captchaImage = payload.properties.captchaImage
+      const captchaId = payload.properties.captchaId
+
+      const promotions = state.promotions.map((item: RewardsExtension.Promotion) => {
+        if (promotionId === item.promotionId) {
+          if (item.captchaStatus !== 'wrongPosition') {
+            item.captchaStatus = 'start'
+          }
+          item.captchaImage = captchaImage
+          item.captchaId = captchaId
+          item.hint = hint
+        }
+        return item
+      })
+
+      state = {
+        ...state,
+        promotions
+      }
+      break
+    }
+    case types.RESET_PROMOTION: {
+      const promotionId = payload.promotionId
+      if (!state.promotions || !promotionId) {
+        break
+      }
+
+      const currentPromotion = getPromotion(promotionId, state.promotions)
+
+      if (!currentPromotion) {
+        break
+      }
+
+      const promotions = state.promotions.map((item: RewardsExtension.Promotion) => {
+        if (currentPromotion.promotionId === item.promotionId) {
+          item.captchaStatus = null
+          item.captchaImage = ''
+          item.captchaId = ''
+          item.hint = ''
+        }
+        return item
+      })
+
+      state = {
+        ...state,
+        promotions
+      }
+      break
+    }
+    case types.DELETE_PROMOTION: {
+      const promotionId = payload.promotionId
+      if (!state.promotions || !promotionId) {
+        break
+      }
+      const currentPromotion = getPromotion(promotionId, state.promotions)
+
+      if (!currentPromotion) {
+        break
+      }
+
+      let promotions = state.promotions.map((item: RewardsExtension.Promotion) => {
+        if (currentPromotion.promotionId === item.promotionId) {
+          item.captchaStatus = null
+          item.status = 4
+        }
+
+        return item
+      })
+
+      state = {
+        ...state,
+        promotions
+      }
+      break
+    }
+    case types.ON_PROMOTION_FINISH: {
+      state = { ...state }
+      let newPromotion: any = {}
+      const result = payload.result
+
+      if (!state.promotions) {
+        break
+      }
+
+      let promotionId = payload.promotion.promotionId
+      newPromotion.promotionId = promotionId
+
+      switch (result) {
+        case 0:
+          newPromotion.captchaStatus = 'finished'
+          newPromotion.captchaImage = ''
+          newPromotion.captchaId = ''
+          newPromotion.hint = ''
+          break
+        case 6:
+          newPromotion.captchaStatus = 'wrongPosition'
+          break
+        default:
+          newPromotion.captchaStatus = 'generalError'
+          break
+      }
+
+      if (state.promotions) {
+        const promotions = updatePromotions(newPromotion, state.promotions)
+
+        state = {
+          ...state,
+          promotions
+        }
+      }
+
+      break
+    }
+  }
+  return state
+}
```

Comment:
This is huge - not being a Typescript / standards expert - at all! but I'd recommend each case is broken out into a single responsibility function.

Reply:
right now we do this is in all reducers. Can you log an issue and can we can refactor all places

Reply:
Raised as https://github.com/brave/brave-browser/issues/6948 :+1: